crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md

```diff
@@ -0,0 +1,134 @@
+# Implicit globals from `types.ModuleType`
+
+## Implicit `ModuleType` globals
+
+All modules are instances of `types.ModuleType`.
+If a name can't be found in any local or global scope, we look it up
+as an attribute on `types.ModuleType` in typeshed
+before deciding that the name is unbound.
+
+```py
+reveal_type(__name__)  # revealed: str
+
+# TODO: infer types from PEP-604 union annotations
+reveal_type(__file__)  # revealed: @Todo
+reveal_type(__loader__)  # revealed: @Todo
+reveal_type(__package__)  # revealed: @Todo
+reveal_type(__spec__)  # revealed: @Todo
+
+# TODO: generics
+reveal_type(__path__)  # revealed: @Todo
+
+# TODO: this should probably be added to typeshed; not sure why it isn't?
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__doc__)
+
+class X:
+    reveal_type(__name__)  # revealed: str
+
+def foo():
+    reveal_type(__name__)  # revealed: str
+```
+
+However, three attributes on `types.ModuleType` are not present as implicit
+module globals; these are excluded:
+
+```py path=unbound_dunders.py
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__getattr__)
+
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__dict__)
+
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__init__)
+```
+
+## Accessed as attributes
```

Comment:
`__class__` is another one defined on `object` that is accessible on `ModuleType` instances, but not as a module global name.

`__module__` is an odd special case in that typeshed says it exists on `object` but it does not actually exist on module objects. Probably not worth modeling this LSP violation, it's fine if we just let it exist on all objects?

Reply:
> `__module__` is an odd special case in that typeshed says it exists on `object` but it does not actually exist on module objects. Probably not worth modeling this LSP violation, it's fine if we just let it exist on all objects?

*Sigh*. `__module__` [really](https://github.com/python/typeshed/pull/8789) [shouldn't](https://github.com/python/typeshed/pull/8787) be listed as an instance attribute on `object`; it's just obviously incorrect. But I'm not going to restart that conversation 🙃

I think this is not worth special-casing in red-knot; we should just accept it as an inaccuracy in type inference that we have to live with unless and until it's fixed in typeshed.